design-proposals/instancetype.kubevirt.io/expand-during-submission.md

```diff
@@ -0,0 +1,301 @@
+Likewise users can explicitly request the `reference` policy to counter a
+different default being defined in the `KubeVirt` `CR`:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: cirros
+spec:
+  instancetype:
+    name: foo
+    policy: reference
+```
+
+Users can also expand an existing `VirtualMachine` by removing any referenced
+revision from a matcher to invoke the expand behaviour, for example starting
+with the following:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: cirros
+spec:
+  instancetype:
+    name: foo
+    revisionName: bar
+```
+
+A user can explicitly ask for the associated instance type to be expanded into
+the `VirtualMachine` by removing `revisionName` and setting `policy` to
+`expand`:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: cirros
+spec:
+  instancetype:
+    name: foo
+    policy: expand
+```
+
+This will in turn expand the instance type into the `VirtualMachine` and remove
+the matcher:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: cirros
+spec:
+  [..]
+```
+
+```go
+// KubeVirtConfiguration holds all kubevirt configurations
+type KubeVirtConfiguration struct {
+[..]
+    // Instancetype configuration
+    Instancetype *InstancetypeConfiguration `json:"instancetype,omitempty"`
+
+    // Preference configuration
+    Preference *PreferenceConfiguration `json:"preference,omitempty"`
+}
+
+type ExpansionPolicy string
+
+const (
+    Expand    ExpansionPolicy = "expand"
+    Reference ExpansionPolicy = "reference"
+)
+
+type InstancetypeConfiguration struct {
+    Policy *ExpansionPolicy `json:"policy,omitempty"`
+}
+
+type PreferenceConfiguration struct {
+    Policy *ExpansionPolicy `json:"policy,omitempty"`
+}
+
+[..]
+
+// InstancetypeMatcher references a instancetype that is used to fill fields in the VMI template.
+type InstancetypeMatcher struct {
+    [..]
+    Policy *ExpansionPolicy `json:"policy,omitempty"`
+}
+
+[..]
+
+// PreferenceMatcher references a set of preference that is used to fill fields in the VMI template.
+type PreferenceMatcher struct {
+    [..]
+    Policy *ExpansionPolicy `json:"policy,omitempty"`
+}
+```
+
+## Scalability
+
+The resulting mutation of the `VirtualMachine` with this proposal will
+cause an additional reconciliation loop to trigger for the VM.
+
+Work should be carried out to ensure that the substantial mutation of the
+`VirtualMachine` during submission doesn't negatively impact the control plane.
+
+## Update/Rollback Compatibility
+
+Existing `VirtualMachines` referencing instance types or preferences will not be
+changed as a result of this behaviour.
+
+There will be no ability to automatically rollback new `VirtualMachines` once
+they have their instance type or preference expanded by this new functionality.
+
+Users will also be unable to resize their `VirtualMachines` by making a singular
+choice of instance type in the future.
```

Comment:
Why not? It is still OK to define instance-typed VM `spec`s, why would it not be ok to modify an existing VM to be instance typed? In properly eventual-consistent systems, the intention of the user is what matters most, not the history of the resource.

Reply:
At present instance types conflict if the VirtualMachine already provides an amount of vCPU or memory.

We have talked about changing this behaviour in https://github.com/kubevirt/community/pull/318 and in doing so could allow a user to resize flat VirtualMachines by selecting an instance type but that might require a follow up design proposal and discussion to work out the details.

Reply:
I do not follow... Assume that I created a flat VM with 7 CPUs. I stopped my VM and then changed my mind. I would like it to be much larger and more standard and edit the VM to use `cx1.8xlarge`. Would KubeVirt attempt to block this? Why?

Reply:
At present yes it would unless you also removed the 7 vCPUs you originally requested within the VirtualMachine.

We just don't allow you to request vCPUs and memory through the VirtualMachineInstanceSpec *and* instance type at the same time.

Reply:
> At present yes it would unless you also removed the 7 vCPUs you originally requested within the VirtualMachine.
> 
> We just don't allow you to request vCPUs and memory through the VirtualMachineInstanceSpec _and_ instance type at the same time.

this is clear and expected. I just don't understand it from "Users will also be unable to resize their `VirtualMachines` by making a singular choice of instance type in the future". Would you agree to rephrase the text?

Reply:
Ah apologies I see your point, I'll rephrase now.